### PR TITLE
New module for setting up Avi Pool

### PR DESCRIPTION
**Pull request by grastogi23**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

avi.avi_pool
##### ANSIBLE VERSION

```
ansible 2.3.0
  config file = /home/grastogi/.ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY

New module for setting up Avi Pool. 

```

```

### CODE REVIEW

**Review by khaltore — APPROVED** [review]
+label shipit

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+            - Reference to server autoscale policy.
+            - It is a reference to an object of type serverautoscalepolicy.
+    capacity_estimation:
+        description:
+            - Inline estimation of capacity of servers.
+        default: False
+    capacity_estimation_ttfb_thresh:
+        description:
+            - The maximum time-to-first-byte of a server.
+        default: 0
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py]: Default not set in argspec.

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+            - Internally set by cloud connector.
+    cloud_ref:
+        description:
+            - It is a reference to an object of type cloud.
+        default: Default-Cloud
+    connection_ramp_duration:
+        description:
+            - Duration for which new connections will be gradually ramped up to a server recently brought online.
+            - Useful for lb algorithms that are least connection based.
+        default: 10
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py]: Default not set in argspec.

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+    created_by:
+        description:
+            - Creator name.
+    default_server_port:
+        description:
+            - Traffic sent to servers will use this destination server port unless overridden by the server's specific port attribute.
+            - The ssl checkbox enables avi to server encryption.
+        default: 80
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py]: Default not set in argspec.

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+            - Comma separated list of domain names which will be used to verify the common names or subject alternative names presented by server certificates.
+            - It is performed only when common name check host_check_enabled is enabled.
+    east_west:
+        description:
+            - Inherited config from virtualservice.
+    enabled:
+        description:
+            - Enable or disable the pool.
+            - Disabling will terminate all open connections and pause health monitors.
+        default: True
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py]: Default not set in argspec.

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+    fail_action:
+        description:
+            - Enable an action - close connection, http redirect, local http response, or backup pool - when a pool failure happens.
+            - By default, a connection will be closed, in case the pool experiences a failure.
+    fewest_tasks_feedback_delay:
+        description:
+            - Periodicity of feedback for fewest tasks server selection algorithm.
+        default: 10
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py]: Default not set in argspec.

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+    graceful_disable_timeout:
+        description:
+            - Used to gracefully disable a server.
+            - Virtual service waits for the specified time before terminating the existing connections  to the servers that are disabled.
+        default: 1
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py]: Default not set in argspec.

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+            - Verify server health by applying one or more health monitors.
+            - Active monitors generate synthetic traffic from each service engine and mark a server up or down based on the response.
+            - The passive monitor listens only to client to server communication.
+            - It raises or lowers the ratio of traffic destined to a server based on successful responses.
+            - It is a reference to an object of type healthmonitor.
+    host_check_enabled:
+        description:
+            - Enable common name check for server certificate.
+            - If enabled and no explicit domain name is specified, avi will use the incoming host header to do the match.
+        default: False
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py]: Default not set in argspec.

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+            - This may alter the expected behavior of the lb method, such as round robin.
+        default: True
+    ipaddrgroup_ref:
+        description:
+            - Use list of servers from ip address group.
+            - It is a reference to an object of type ipaddrgroup.
+    lb_algorithm:
+        description:
+            - The load balancing algorithm will pick a server within the pool's list of available servers.
+        default: 1
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py]: Default not set in argspec.

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+    lb_algorithm_consistent_hash_hdr:
+        description:
+            - Http header name to be used for the hash key.
+    lb_algorithm_hash:
+        description:
+            - Criteria used as a key for determining the hash between the client and  server.
+        default: 1
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py]: Default not set in argspec.

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+    max_concurrent_connections_per_server:
+        description:
+            - The maximum number of concurrent connections allowed to each server within the pool.
+        default: 0
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py]: Default not set in argspec.

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+    max_conn_rate_per_server:
+        description:
+            - Rate limit connections to each server.
+    name:
+        description:
+            - The name of the pool.
+        required: true
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py:171]: Required not set in argspec

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+    networks:
+        description:
+            - (internal-use) networks designated as containing servers for this pool.
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py:174]: If this is for internal use does it need to be in the module?

File: lib/ansible/modules/network/avi/avi_pool.py [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+            - Manually select the networks and subnets used to provide reachability to the pool's servers.
+            - Specify the subnet using the following syntax  10-1-1-0/24.
+            - Use static routes in vrf configuration when pool servers are not directly connected butroutable from the service engine.
+    prst_hdr_name:
+        description:
+            - Header name for custom header persistence.
+    request_queue_depth:
+        description:
+            - Minimum number of requests to be queued when pool is full.
+        default: 128
```
Comment (gundalow) [inline lib/ansible/modules/network/avi/avi_pool.py]: Default not set in argspec.

**Review by gundalow — CHANGES_REQUESTED** [review]
I've not been through everything, though please ensure if in the documentation you specifuy:
* a default
* required: True

That this is also reflected in `argspec`

File: lib/ansible/modules/network/avi/avi_pool.py (hunk 11 above)
Comment (grastogi23) [reply lib/ansible/modules/network/avi/avi_pool.py:171]: fixed

File: lib/ansible/modules/network/avi/avi_pool.py (hunk 12 above)
Comment (grastogi23) [reply lib/ansible/modules/network/avi/avi_pool.py:174]: This information is still returned in the API. It is just not editable. In next version of Avi we will clean it up then I would be updating it to deprecated. However, till our next release happens we need to keep it in the object.

**Comment by grastogi23** [comment]
Fixed all other items.

**Review by gundalow — APPROVED** [review]

**Comment by ericsysmin** [comment]
shipit

**Comment by ericsysmin** [comment]
+shipit

**Comment by khaltore** [comment]
shipit

**Comment by ericsysmin** [comment]
shipit

**Comment by ericsysmin** [comment]
shipit

**Pull request merged**